internal/venice-common/src/main/java/com/linkedin/venice/utils/DiskUsage.java

```diff
@@ -85,8 +127,130 @@ public boolean isDiskFull(long bytesWritten) {
-        + " bytes free.  " + "Disk is marked 'full' when there are less than " + freeSpaceBytesRequired
-        + " bytes free.";
+    StringBuilder sb = new StringBuilder();
+    if (databaseSizeLimiter != null) {
+      sb.append(databaseSizeLimiter.getDatabaseStatus()).append(", ");
+    }
+    sb.append("Disk at path: ")
+        .append(diskPath)
+        .append(" has ")
+        .append(totalSpaceBytes)
+        .append(" total bytes and ")
+        .append(freeSpaceBytes)
+        .append(" bytes free. ")
+        .append("Disk is marked 'full' when there are less than ")
+        .append(freeSpaceBytesRequired)
+        .append(" bytes free.");
+    return sb.toString();
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (databaseSizeLimiter != null) {
+      databaseSizeLimiter.shutdown();
+    }
+  }
+
+  /**
+   * This class is used to measure whether the current database size exceeds the limit or not.
+   * It will measure the size after writing a configured amount of bytes, and the measurement
+   * won't be very accurate since the measurement is independent of the database write operation,
+   * and it was designed in this way to minimize the overhead of this class in the write path.
+   */
+  public static class DatabaseSizeLimiter {
+    private static final Logger LOGGER = LogManager.getLogger(DatabaseSizeLimiter.class);
+    public static final long DEFAULT_DATABASE_SIZE_MEASURE_INTERVAL = 256 * 1024 * 1024; // 256MB
+
+    private final String databasePath;
+    private final long databaseSizeLimit;
+    private final long databaseSizeMeasurementInterval;
+    private final AtomicLong currentDatabaseSize = new AtomicLong();
+    private final Supplier<Long> databaseSizeSupplier;
+    private final AtomicLong accumulatedByteSize = new AtomicLong();
+    private final AtomicBoolean hasOngoingMeasurement = new AtomicBoolean(false);
+    private final ExecutorService databaseSizeMeasurementExecutor =
+        Executors.newSingleThreadExecutor(new DefaultThreadFactory("Database_Size_Measurement"));
+    private boolean exceedDatabaseSizeLimit = false;
+
+    public DatabaseSizeLimiter(String databasePath, long databaseSizeLimit, long databaseSizeMeasurementInterval) {
+      this(
+          databasePath,
+          databaseSizeLimit,
+          databaseSizeMeasurementInterval,
+          () -> FileUtils.sizeOf(new File(databasePath)));
+    }
+
+    // For testing
+    public DatabaseSizeLimiter(
+        String databasePath,
+        long databaseSizeLimit,
+        long databaseSizeMeasurementInterval,
+        Supplier<Long> databaseSizeSupplier) {
+      this.databasePath = databasePath;
+      this.databaseSizeLimit = databaseSizeLimit;
+      this.databaseSizeMeasurementInterval = databaseSizeMeasurementInterval;
+      this.databaseSizeSupplier = databaseSizeSupplier;
+    }
+
+    public boolean hitDatabaseSizeLimit(long bytesWritten) {
```

Comment:
What if we are writing many duplicate keys, it will count each key as a separate one and add them up?